plugins/lines/prism-lines.js

```diff
@@ -0,0 +1,130 @@
+(function () {
+
+	if (typeof self === 'undefined' || !self.Prism || !self.document) {
+		return;
+	}
+
+	/**
+	 * Returns style declarations for the element
+	 * @param {Element} element
+	 */
+	Prism.hooks.add('before-insert', function (env) {
+
+		// works only for <code> wrapped inside <pre> (not inline)
+		var pre = env.element.parentNode;
+		if (!pre || !/pre/i.test(pre.nodeName)) {
+			return;
+		}
+
+		// we are not welcome here :(
+		if (pre.className.includes('nolines')) {
+			return;
+		}
+
+		// support for line-numbers behaviour
+		var dataStart = 1;
+		if (pre.hasAttribute('data-start')) {
+			dataStart = parseInt(pre.getAttribute('data-start'), 10);
+			if (isNaN(dataStart)) {
+				dataStart = 1;
+			}
+			pre.style.setProperty('--counter-start', dataStart - 1);
+                }
+
+		function createLine(lineNumber, prefix) {
+			var line = document.createElement('div');
+			var anchor = document.createElement('a');
+			if (prefix) {
+				var target = prefix + '.' + lineNumber;
+				line.id = target;
+				anchor.href = '#' + target;
+			}
+			line.setAttribute('class', 'line');
+			line.appendChild(anchor);
+			return line;
+		}
+
+		function splitNode(node) {
+			if (node.hasChildNodes()) {
+				// so we've got a kid! :)
+				var p1 = document.createElement(node.nodeName);
+				var p2 = document.createElement(node.nodeName);
+				if (node.hasAttributes()) {
+					for (i = 0; i < node.attributes.length ; i++) {
```

Comment:
Today I learned something I now don;t like about JavaScript, thanks!  It's counter-intuitive, but you are right about the scope.

Reply:
That's just how it works. You basically said that `i` is a global variable and JS just did as it was told. 

Well, this is a problem that (almost) every dynamically typed language has since variables can be (re-)defined at runtime. 

In JavaScript, everything outside of `var`s inside functions (different behavior for `let`/`const`) is a global variable. `Array`, `Object`, `RegExp` are all global variables. You can even delete or replace them if you wanted to. (Please don't.)
This might seem horrible at first (and in many ways, it is) but this also really powerful because, in JS, you can bend everything to your will. (This is also what enables polyfills.)

Well, modern language features such as `let` and `const` are a lot stricter tho.